File: Adhiti_Shaili/Day 10/Search 2D matrix.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        if(matrix.size()==0)
        {
            return false;
        }
        
        int m=matrix.size();
        int n=matrix[0].size();
        
        if(m==1&&n==1)//&&matrix[0][0]==target)
        {
            if(matrix[0][0]==target)
            {
                return true;
            }
            else
            {
                return false;
            }
        }
        
        int low=0;
        int high=m-1;
        
        int row=0;
       // int col=0;
        
        while(low<high)//
        {
            int mid=(low+high)/2;
            if(matrix[mid][n-1]==target)
            {
                // row=mid;
                // break;
                return true;
            }
            else if(matrix[mid][n-1]<target)
            {
                low=mid+1;
            }
            else
            {
                high=mid;
            }
            row=low;
        }
        
        low=0;
        high=n-1;
        while(low<=high)//
        {
            int mid=(low+high)/2;
            if(matrix[row][mid]==target)
            {
                return true;
            }
            else if(matrix[row][mid]<target)
            {
                low=mid+1;
            }
            else
            {
                high=mid-1;
            }
            //row=low;
        }
        return false;
        
    }
};

File: sanket-dalvi/Day10/search2DMatrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if len(matrix) == 0: return False
        m = len(matrix)
        n = len(matrix[0])
        def getInd(i):
            return (i//n, i%n)
        l = 0
        h = m*n - 1
        while l <= h:
            mid = l + (h-l)//2
            i, j = getInd(mid)
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] < target:
                l = mid + 1
            else:
                h = mid - 1
        return False

# Test
print(Solution().searchMatrix(matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,50]], target = 34))